specifyweb/specify/tree_views.py

```diff
@@ -342,4 +343,361 @@ def perm_target(tree):
+    'INFRAORDER': 120,
+    'PARVORDER': 125,
+    'SUPERFAMILY': 130,
+    'FAMILY': 140,
+    'SUBFAMILY': 150,
+    'TRIBE': 160,
+    'SUBTRIBE': 170,
+    'GENUS': 180,
+    'SUBGENUS': 190,
+    'SECTION': 200,
+    'SUBSECTION': 210,
+    'SPECIES': 220,
+    'SUBSPECIES': 230,
+    'VARIETY': 240,
+    'SUBVARIETY': 250,
+    'FORMA': 260,
+    'SUBFORMA': 270
+}
+GEOGRAPHY_RANKS = {
+    'CONTINENT': 100,
+    'COUNTRY': 200,
+    'STATE': 300,
+    'COUNTY': 400
+}
+STORAGE_RANKS = {
+    'BUILDING': 100,
+    'COLLECTION': 150,
+    'ROOM': 200,
+    'AISLE': 250,
+    'CABINET': 300,
+    'SHELF': 350,
+    'BOX': 400,
+    'RACK': 450,
+    'VIAL': 500
+}
+GEOLOGIC_TIME_PERIOD_RANKS = {
+    'ERA': 100,
+    'PERIOD': 200,
+    'EPOCH': 300,
+    'AGE': 400
+}
+LITHO_STRAT_RANKS = {
+    'SUPERGROUP': 100,
+    'GROUP': 200,
+    'FORMATION': 300,
+    'MEMBER': 400,
+    'BED': 500
+}
+
+DEFAULT_RANK_INCREMENT = 100
+TAXON_RANK_INCREMENT = 10
+GEOGRAPHY_RANK_INCREMENT = DEFAULT_RANK_INCREMENT
+STORAGE_RANK_INCREMENT = 50
+GEOLOGIC_TIME_PERIOD_INCREMENT = DEFAULT_RANK_INCREMENT
+LITHO_STRAT_INCREMENT = DEFAULT_RANK_INCREMENT
+
+TREE_RANKS_MAPPING = {
+    'taxon': (TAXON_RANKS, TAXON_RANK_INCREMENT),
+    'geography': (GEOGRAPHY_RANKS, GEOGRAPHY_RANK_INCREMENT),
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+@openapi.schema({
+    "post": {
+        "parameters": [
+            {
+                "name": "newRankName",
+                "in": "formData",
+                "required": True,
+                "schema": {
+                    "type": "string"
+                },
+                "description": "The name of the new rank to add"
+            },
+            {
+                "name": "targetRankName",
+                "in": "formData",
+                "required": True,
+                "schema": {
+                    "type": "string"
+                },
+                "description": "The name of the parent rank to add the new rank to (use 'root' to add to the front)"
+            },
+            {
+                "name": "treeType",
+                "in": "formData",
+                "required": True,
+                "schema": {
+                    "type": "string",
+                    "enum": ["taxon", "geography", "storage", "geologicTimePeriod", "lithoStrat"]
+                },
+                "description": "The type of the tree (taxon, geography, or storage)"
+            },
+            {
+                "name": "newRankTitle",
+                "in": "formData",
+                "required": False,
+                "schema": {
+                    "type": "string"
+                },
+                "description": "The title of the rank to add (defaults to the name)"
+            },
+            {
+                "name": "useDefaultRankIDs",
```

Comment:
what is the use case for the front-end to pass this parameter? seems like your logic for determining new rank id is robust enough already without requiring additional input from the front-end

Reply:
I could go either way with this.  I wanted to be able to handle the case where the client wants to use a default name, but not the default rank id.  It's also not a required field, it's just there to provide clarity to the fact that there are default ranks to the client.